components/resource-creation/ResourcePanel.tsx

```diff
@@ -1,16 +1,49 @@
-import { Center, Divider, Text } from '@mantine/core';
-import { Suspense } from 'react';
-import { useRecoilValue } from 'recoil';
+import { Button, Center, Divider, Text } from '@mantine/core';
+import { v4 as uuidv4 } from 'uuid';
+import { IconAlertCircle, IconCodePlus } from '@tabler/icons';
+import { Suspense, useState } from 'react';
+import { useRecoilState, useRecoilValue } from 'recoil';
 import { Loader } from 'tabler-icons-react';
 import { patientTestCaseState } from '../../state/atoms/patientTestCase';
 import { selectedPatientState } from '../../state/atoms/selectedPatient';
 import { getPatientNameString } from '../../util/fhir';
+import produce from 'immer';
+import CodeEditorModal from '../modals/CodeEditorModal';
 import ResourceDisplay from './ResourceDisplay';
 import ResourceSelection from './ResourceSelection';
+import { showNotification } from '@mantine/notifications';
 
 export default function ResourcePanel() {
   const selectedPatient = useRecoilValue(selectedPatientState);
-  const currentPatients = useRecoilValue(patientTestCaseState);
+  const [currentPatients, setCurrentPatients] = useRecoilState(patientTestCaseState);
+  const [isNewResourceModalOpen, setIsNewResourceModalOpen] = useState(false);
+
+  const createNewResource = (val: string) => {
+    // TODO: Validate the incoming JSON as FHIR
+    const newResource = JSON.parse(val.trim());
+    // Create a new state object using immer without needing to shallow clone the entire previous object
+    if (selectedPatient) {
+      if (!newResource.id) {
+        newResource.id = uuidv4();
+      }
+      const resourceIndex = currentPatients[selectedPatient].resources.findIndex(r => r.id === newResource.id);
+      if (resourceIndex >= 0) {
+        showNotification({
+          id: 'failed-upload',
+          icon: <IconAlertCircle />,
+          title: 'Resource Creation Failed',
+          message: `Resource ID ${newResource.id} is used for an existing resource for this patient.`,
+          color: 'red'
+        });
+      } else {
+        const nextResourceState = produce(currentPatients, draftState => {
+          draftState[selectedPatient].resources.push(newResource);
+        });
+        setCurrentPatients(nextResourceState);
+      }
+    }
+    setIsNewResourceModalOpen(false);
```

Comment:
This logic always closes the modal even in the error case, making the user lose their progress if all they needed to do was change the ID. Maybe we should move this line into the else case above so the modal closes only if the state is actually about to be updated? Otherwise they can correct their error and save again